File: app/src/main/kotlin/com/lelemusic/model/LyricParsers.kt
package com.lelemusic.model

/**
 * LRC / SRT / 纯文本歌词解析器（纯 Kotlin，无 Android 依赖，可在 JVM 单测）。
 *
 * 解析原则：**只提取「能滚动高亮的同步行」**，元信息行（`[ti:]`/`[ar:]`/SRT 序号行等）
 * 一律丢弃；纯文本（无任何时间戳）退化为整段静态歌词，[Lyric.synced] = false。
 */
object LyricParsers {

    // -----------------------------------------------------------------------
    // LRC
    // -----------------------------------------------------------------------

    /** `[mm:ss.xx]` / `[mm:ss:xxx]` 三种小数分隔都兼容；一行可挂多个时间戳 */
    private val LRC_TAG = Regex("\\[(\\d{1,3}):(\\d{1,2})(?:[.:](\\d{1,3}))?]")

    /**
     * 解析 LRC 文本。
     *
     * - `[mm:ss.xx]` 的小数是**厘秒**（2 位），换算 ms = ×10；
     *   若出现 3 位则按毫秒原值处理（部分客户端写法）。
     * - 同一行多个时间戳 → 拆成多行（同一文本重复输出）。
     */
    fun parseLrc(raw: String): Lyric {
        val out = ArrayList<LyricLine>(32)
        var foundTimestamp = false
        for (lineRaw in raw.lineSequence()) {
            val tags = LRC_TAG.findAll(lineRaw).toList()
            if (tags.isEmpty()) continue
            foundTimestamp = true
            val text = lineRaw.substring(tags.last().range.last + 1).trim()
            if (text.isBlank()) continue
            for (tag in tags) {
                val min = tag.groupValues[1].toLongOrNull() ?: 0L
                val sec = tag.groupValues[2].toLongOrNull() ?: 0L
                val frac = tag.groupValues[3]
                val ms = min * 60_000L + sec * 1_000L + fracMs(frac)
                out.add(LyricLine(timeMs = ms, text = text))
            }
        }
        if (!foundTimestamp || out.isEmpty()) return fromPlain(raw)
        out.sortBy { it.timeMs }
        return Lyric(lines = out, synced = true)
    }

    // -----------------------------------------------------------------------
    // SRT
    // -----------------------------------------------------------------------

    private val SRT_TIME = Regex(
        "(\\d{1,3}):(\\d{1,2}):(\\d{1,2})[,.]\\d{1,3}\\s*-->\\s*" +
            "(\\d{1,3}):(\\d{1,2}):(\\d{1,2})[,.]\\d{1,3}"
    )

    /**
     * 解析 SRT 字幕文本。
     *
     * 每个块：可选序号行 → `hh:mm:ss,mmm --> ...` 时间行 → 字幕文本行。
     * 取块的**起始时间**作为 [LyricLine.timeMs]，文本行用 `\n` 连接（多行字幕）。
     */
    fun parseSrt(raw: String): Lyric {
        val out = ArrayList<LyricLine>(32)
        for (block in raw.split(Regex("\\R\\s*\\R"))) {
            if (block.isBlank()) continue
            val lines = block.trim().lines()
            val timeIndex = lines.indexOfFirst { SRT_TIME.containsMatchIn(it) }
            if (timeIndex < 0) continue
            val match = SRT_TIME.find(lines[timeIndex]) ?: continue
            val startMs = match.groupValues[1].toLongOrNull()?.let { h ->
                val m = match.groupValues[2].toLongOrNull() ?: 0L
                val s = match.groupValues[3].toLongOrNull() ?: 0L
                h * 3_600_000L + m * 60_000L + s * 1_000L
            } ?: continue
            val text = lines
                .drop(timeIndex + 1)
                .map { it.trim() }
                .filter { it.isNotBlank() }
                .joinToString("\n")
            if (text.isBlank()) continue
            out.add(LyricLine(timeMs = startMs, text = text))
        }
        if (out.isEmpty()) return fromPlain(raw)
        out.sortBy { it.timeMs }
        return Lyric(lines = out, synced = true)
    }

    // -----------------------------------------------------------------------
    // 统一入口 + 纯文本兜底
    // -----------------------------------------------------------------------

    /**
     * 按文件名后缀决定解析器；无法识别时智能兜底：
     * 内容含 `-->` 视为 SRT，否则按 LRC 处理（LRC 解析器对纯文本自然退回静态歌词）。
     *
     * @return 永不返回 null：解析结果最差是 [Lyric.isEmpty] 为 true 的静态歌词
     */
    fun parseFile(fileName: String?, content: String): Lyric {
        val lower = fileName?.lowercase().orEmpty()
        return when {
            lower.endsWith(".srt") -> parseSrt(content)
            content.contains("-->") -> parseSrt(content)
            else -> parseLrc(content)
        }
    }

    /** 无时间轴文本 → 静态歌词：逐行保留非空行，[Lyric.synced] = false */
    fun fromPlain(raw: String): Lyric {
        val lines = raw.lineSequence()
            .map { it.trimEnd() }
            .filter { it.isNotBlank() }
            .map { LyricLine(timeMs = -1L, text = it) }
            .toList()
        return Lyric(lines = lines, synced = false)
    }

    private fun fracMs(frac: String): Long = when (frac.length) {
        3 -> frac.toLongOrNull() ?: 0L
        2 -> (frac.toLongOrNull() ?: 0L) * 10L
        1 -> (frac.toLongOrNull() ?: 0L) * 100L
        else -> 0L
    }
}

File: ios/LeleMusic/Data/LyricParsers.swift
import Foundation

// MARK: - 音源接口地址注册表（对齐 Android ResolverEndpoints）

enum ResolverEndpoints {
    private static let lock = NSLock()
    private static var defaults: [String: String] = [:]
    private static var overrides: [String: String] = [:]

    /// resolver 登记内置默认地址（重复注册无害，后者覆盖）
    static func register(_ strategyId: String, _ defaultUrl: String) {
        lock.lock()
        defer { lock.unlock() }
        guard !defaultUrl.isEmpty else { return }
        defaults[strategyId] = defaultUrl
    }

    /// 回灌用户覆盖（UserDefaults 持久化）
    static func load() {
        lock.lock()
        defer { lock.unlock() }
        let dict = UserDefaults.standard.dictionary(forKey: "endpoint_overrides") as? [String: String] ?? [:]
        overrides = dict
    }

    /// 当前应使用的地址：用户覆盖 > 默认
    static func current(_ strategyId: String) -> String? {
        lock.lock()
        defer { lock.unlock() }
        return overrides[strategyId] ?? defaults[strategyId]
    }

    static func defaultOf(_ strategyId: String) -> String? {
        lock.lock()
        defer { lock.unlock() }
        return defaults[strategyId]
    }

    static func isOverridden(_ strategyId: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return overrides.keys.contains(strategyId)
    }

    static func setOverride(_ strategyId: String, _ url: String) {
        lock.lock()
        overrides[strategyId] = url.isEmpty ? nil : url
        lock.unlock()
        persist()
    }

    static func clearOverride(_ strategyId: String) {
        lock.lock()
        overrides.removeValue(forKey: strategyId)
        lock.unlock()
        persist()
    }

    private static func persist() {
        lock.lock()
        let dict = overrides
        lock.unlock()
        UserDefaults.standard.set(dict, forKey: "endpoint_overrides")
    }
}

// MARK: - 歌词解析器（纯 Swift 移植 Android LyricParsers）

struct LyricLine: Identifiable, Hashable {
    /// 毫秒；纯文本歌词统一为 -1
    var timeMs: Int
    var text: String
    var id: Int { timeMs ^ text.hashValue }
}

struct Lyric {
    var lines: [LyricLine]
    var synced: Bool

    var isEmpty: Bool { !lines.contains { !$0.text.trimmingCharacters(in: .whitespaces).isEmpty } }

    /// 播放到 ms 时应高亮的行下标；空/未同步返回 -1
    func indexAt(_ ms: Int) -> Int {
        guard !lines.isEmpty, synced else { return -1 }
        var index = 0
        while index < lines.count - 1 && lines[index + 1].timeMs <= ms { index += 1 }
        return index
    }
}

enum LyricParsers {
    /// `[mm:ss.xx]` / `[mm:ss:xxx]` 兼容；一行可挂多个时间戳
    private static let lrcTag = try! NSRegularExpression(pattern: #"\[(\d{1,3}):(\d{1,2})(?:[.:](\d{1,3}))?\]"#)

    static func parseLrc(_ raw: String) -> Lyric {
        var out: [LyricLine] = []
        var foundTimestamp = false
        for lineRaw in raw.components(separatedBy: .newlines) {
            let ns = lineRaw as NSString
            let matches = lrcTag.matches(in: lineRaw, range: NSRange(location: 0, length: ns.length))
            if matches.isEmpty { continue }
            foundTimestamp = true
            guard let last = matches.last else { continue }
            let text = ns.substring(from: last.range.upperBound).trimmingCharacters(in: .whitespaces)
            if text.isEmpty { continue }
            for m in matches {
                let minute = Int(ns.substring(with: m.range(at: 1))) ?? 0
                let second = Int(ns.substring(with: m.range(at: 2))) ?? 0
                var frac = ""
                if m.range(at: 3).location != NSNotFound {
                    frac = ns.substring(with: m.range(at: 3))
                }
                let ms = minute * 60_000 + second * 1_000 + fracMs(frac)
                out.append(LyricLine(timeMs: ms, text: text))
            }
        }
        if !foundTimestamp || out.isEmpty { return fromPlain(raw) }
        out.sort { $0.timeMs < $1.timeMs }
        return Lyric(lines: out, synced: true)
    }

    /// 无时间轴文本 → 静态歌词
    static func fromPlain(_ raw: String) -> Lyric {
        let lines = raw
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines.union(.init(charactersIn: "\r"))) }
            .filter { !$0.isEmpty }
            .map { LyricLine(timeMs: -1, text: $0) }
        return Lyric(lines: lines, synced: false)
    }

    /// 内容含 `-->` 视为 SRT（简化：MVP 只保留 LRC 主链路，SRT 走纯文本）
    static func parseFile(fileName: String?, content: String) -> Lyric {
        parseLrc(content)
    }

    private static func fracMs(_ frac: String) -> Int {
        guard let v = Int(frac) else { return 0 }
        switch frac.count {
        case 3: return v
        case 2: return v * 10
        case 1: return v * 100
        default: return 0
        }
    }
}

// MARK: - 歌词加载结果三态（对齐 LyricResult）

enum LyricResult {
    case success(Lyric)
    case noLyric
    case error(String, String)
}
